eventauth/eventauth.go

```diff
@@ -362,3 +368,199 @@ func (e *eventAllower) commonChecks(event Event) error {
+func (m *membershipAllower) membershipAllowedSelf() error {
+	if m.newMember.Membership == "join" {
+		// A user that is not in the room is allowed to join if the room
+		// join rules are "public".
+		if m.oldMember.Membership == "leave" && m.joinRule.JoinRule == "public" {
+			return nil
+		}
+		// An invited user is allowed to join if the join rules are "public"
+		if m.oldMember.Membership == "invite" && m.joinRule.JoinRule == "public" {
+			return nil
+		}
+		// An invited user is allowed to join if the join rules are "invite"
+		if m.oldMember.Membership == "invite" && m.joinRule.JoinRule == "invite" {
+			return nil
+		}
+		// A joined user is allowed to update their join.
+		if m.oldMember.Membership == "join" {
+			return nil
+		}
+	}
+	if m.newMember.Membership == "leave" {
+		// A joined user is allowed to leave the room.
+		if m.oldMember.Membership == "join" {
+			return nil
+		}
+		// An invited user is allowed to reject an invite.
+		if m.oldMember.Membership == "invite" {
+			return nil
+		}
+	}
+	return m.membershipFailed()
+}
+
+// membershipAllowedSelf determines if the user is allowed to change the membership of another user.
+func (m *membershipAllower) membershipAllowedOther() error {
+	senderLevel := m.powerLevels.userLevel(m.senderID)
+	targetLevel := m.powerLevels.userLevel(m.targetID)
+
+	// You may only modify the membership of another user if you are in the room.
+	if m.senderMember.Membership == "join" {
+		if m.newMember.Membership == "ban" {
+			// A user may ban another user if their level is high enough
+			if senderLevel >= m.powerLevels.banLevel &&
+				senderLevel > targetLevel {
+				return nil
+			}
+		}
+		if m.newMember.Membership == "leave" {
+			// A user may unban another user if their level is high enough.
+			if m.oldMember.Membership == "ban" && senderLevel >= m.powerLevels.banLevel {
+				return nil
+			}
+			// A user may kick another user if their level is high enough.
+			if m.oldMember.Membership != "ban" &&
+				senderLevel >= m.powerLevels.kickLevel &&
+				senderLevel > targetLevel {
+				return nil
+			}
+		}
+		if m.newMember.Membership == "invite" {
+			// A user may invite another user if the user has left the room.
+			// and their level is high enough.
+			if m.oldMember.Membership == "leave" && senderLevel >= m.powerLevels.inviteLevel {
+				return nil
+			}
+			// A user may re-invite a user.
+			if m.oldMember.Membership == "invite" && senderLevel >= m.powerLevels.inviteLevel {
+				return nil
+			}
+		}
+	}
+
+	return m.membershipFailed()
+}
+
+// membershipFailed returns a error explaining why the membership change was disallowed.
+func (m *membershipAllower) membershipFailed() error {
+	if m.senderID == m.targetID {
+		return errorf(
+			"%q is not allowed to change their membership from %q to %q",
+			m.targetID, m.oldMember.Membership, m.newMember.Membership,
+		)
+	}
+
+	if m.senderMember.Membership != "join" {
+		return errorf("sender %q is not in the room", m.senderID)
```

Comment:
I feel it would be polite to tell the client what their membership actually *is*.